BLE_GattServer_FOTAService/source/BlockDeviceFOTAEventHandler.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Mbed-OS Microcontroller Library
+ * Copyright (c) 2020-2021 Embedded Planet
+ * Copyright (c) 2020-2021 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+
+#include "BlockDeviceFOTAEventHandler.h"
+
+#include "BlockDevice.h"
+#include "mbed-trace/mbed_trace.h"
+
+#define TRACE_GROUP "FOTA"
+
+BlockDeviceFOTAEventHandler::BlockDeviceFOTAEventHandler(mbed::BlockDevice &bd, events::EventQueue &eq) :
+                                                         _bd(bd), _eq(eq)
+{
+}
+
+BlockDeviceFOTAEventHandler::~BlockDeviceFOTAEventHandler()
+{
+    delete _bd_eraser;
+}
+
+FOTAService::StatusCode_t BlockDeviceFOTAEventHandler::on_binary_stream_written(
+        FOTAService &fota_service,
+        mbed::Span<const uint8_t> buffer)
+{
+    tr_info("BSC written: programming %d bytes at address %" PRIu64 "",
+            buffer.size(), _addr);
+
+    int error = _bd.program(buffer.data(), _addr, buffer.size());
+    if (error) {
+        tr_error("Programming block device failed: 0x%x", error);
+        return FOTAService::FOTA_STATUS_MEMORY_ERROR;
+    }
+
+    _addr += buffer.size();
+
+    return FOTAService::FOTA_STATUS_OK;
+}
+
+GattAuthCallbackReply_t BlockDeviceFOTAEventHandler::on_control_written(
+        FOTAService &fota_service,
+        mbed::Span<const uint8_t> buffer)
+{
+    _fota_service = &fota_service;
+
+    switch (buffer[0]) {
+        case FOTAService::FOTA_NO_OP:
+        {
+            break;
+        }
+
+        case FOTAService::FOTA_START:
+        {
+            tr_info("Starting FOTA session");
+            fota_service.start_fota_session();
+
+            /* We must erase the update block device before accepting BSC writes
+             * Therefore, set the status characteristic to XOFF to initiate a "delayed start"
+             */
+            fota_service.set_xoff();
+
+            /* Delete previously allocated eraser */
+            delete _bd_eraser;
```

Comment:
how do you know it's safe to delete?

Reply:
Are you worried that it might be a null pointer? Deleting a `nullptr` has no effect, so the code above is equivalent to:

```C++
            /* Delete previously allocated eraser */
            if (_bd_eraser) {
                delete _bd_eraser;
            }
```

Reply:
I am not worried it might be null, I'm worried it might be still executing

Reply:
Commit e698cc23e87dea28fe54677142d743753770bfe1 resolves this.